homeassistant/components/ecowitt/__init__.py

```diff
@@ -0,0 +1,178 @@
+    CONF_UNIT_LIGHTNING,
+    CONF_UNIT_RAIN,
+    CONF_UNIT_WIND,
+    CONF_UNIT_WINDCHILL,
+    DATA_MODEL,
+    DATA_PASSKEY,
+    DATA_STATIONTYPE,
+    DOMAIN,
+    ECOWITT_PLATFORMS,
+    SIGNAL_NEW_SENSOR,
+    SIGNAL_REMOVE_ENTITIES,
+    SIGNAL_UPDATE,
+    W_TYPE_HYBRID,
+    W_TYPE_NEW,
+    W_TYPE_OLD,
+)
+
+NOTIFICATION_ID = DOMAIN
+NOTIFICATION_TITLE = "Ecowitt config migrated"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class EcowittData:
+    """Internal data class for the ecowitt integration."""
+
+    client: type[EcoWittListener]
+    registered_devices: list[str]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up the Ecowitt component from UI."""
+
+    if not entry.options:
+        options = {
+            CONF_UNIT_BARO: CONF_UNIT_SYSTEM_METRIC,
+            CONF_UNIT_WIND: CONF_UNIT_SYSTEM_IMPERIAL,
+            CONF_UNIT_RAIN: CONF_UNIT_SYSTEM_IMPERIAL,
+            CONF_UNIT_LIGHTNING: CONF_UNIT_SYSTEM_IMPERIAL,
+            CONF_UNIT_WINDCHILL: W_TYPE_HYBRID,
+        }
+        hass.config_entries.async_update_entry(entry, options=options)
+
+    # setup the base connection
+    data = EcowittData(EcoWittListener(port=entry.data[CONF_PORT]), [])
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = data
+
+    if entry.options[CONF_UNIT_WINDCHILL] == W_TYPE_OLD:
+        data.client.set_windchill(WINDCHILL_OLD)
+    if entry.options[CONF_UNIT_WINDCHILL] == W_TYPE_NEW:
+        data.client.set_windchill(WINDCHILL_NEW)
+    if entry.options[CONF_UNIT_WINDCHILL] == W_TYPE_HYBRID:
+        data.client.set_windchill(WINDCHILL_HYBRID)
+
+    hass.loop.create_task(data.client.listen())
+    await data.client.wait_for_valid_data()
+
+    hass.config_entries.async_setup_platforms(entry, ECOWITT_PLATFORMS)
+
+    def _new_sensor_callback():
+        """Prepare callback for a new sensor addition."""
+        _LOGGER.debug("New sensor callback triggered")
+        for platform in ECOWITT_PLATFORMS:
+            async_dispatcher_send(
+                hass, SIGNAL_NEW_SENSOR.format(platform, entry.entry_id)
+            )
+
+    # set the callback so we know we have new sensors
+    data.client.new_sensor_cb = _new_sensor_callback
+
+    async def _async_ecowitt_update_cb(weather_data):
+        """Primary update callback called from pyecowitt."""
+        _LOGGER.debug("Primary update callback triggered")
+        async_dispatcher_send(hass, SIGNAL_UPDATE.format(entry.entry_id))
+
+    # this is part of the base async_setup_entry
+    data.client.register_listener(_async_ecowitt_update_cb)
+    entry.async_on_unload(entry.add_update_listener(update_listener))
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+
+    data = hass.data[DOMAIN][entry.entry_id]
+    await data.client.stop()
+
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        entry, ECOWITT_PLATFORMS
```

Comment:
No need to set the `ECOWITT_` prefix:
```suggestion
        entry, PLATFORMS
```